tsdb/engine/tsm1/engine.go

```diff
@@ -228,7 +258,8 @@ func (e *DevEngine) Begin(writable bool) (tsdb.Tx, error) {
 
 func (e *DevEngine) WriteTo(w io.Writer) (n int64, err error) { panic("not implemented") }
 
-func (e *DevEngine) writeSnapshot() error {
+// WriteSnapshot will snapshot the cache and write a new TSM file with its contents, releasing the snapshot when done.
+func (e *DevEngine) WriteSnapshot(async bool) error {
```

Comment:
Here's another situation. Say snapshots fail to save for some underlying system reason. Could be file permissions, full HD, or whatever.

We continue to take writes until we fill up memory then reject them (because everything is backing up in the WAL). The user restarts the process, we take a long time to reload the cache (so now because we didn't error hard early, it takes a longer time to recover), and we immediately fill up the cache to the point where we can no longer accept writes again.

There are many situations in which it's more user friendly to let the process go down so the user can quickly look into it. I'm willing to be that there are many people out there that are not tailing their logs for errors. It's even worse because who knows how many errors we may be throwing out? We don't have levels clearly defined.

It's a poor user experience to throw up our hands and say "well you should have had better monitoring".

Continuing to answer queries, but throw writes on the floor because there are underlying system problems is a very poor experience. I still maintain that a system level error on writes should stop everything cold. Otherwise users will complain that we lost their data because we kept the database running (even if we returned 500s on the writes). And what about UDP? There's no way for the client to know.